### PR TITLE
[SPARK-12612][PROJECT-INFRA] Add missing Hadoop profiles to dev/run-tests-*.py scripts and dev/deps

### PR DESCRIPTION
**Pull request by JoshRosen**
There are a couple of places in the `dev/run-tests-*.py` scripts which deal with Hadoop profiles, but the set of profiles that they handle does not include all Hadoop profiles defined in our POM. Similarly, the `hadoop-2.2` and `hadoop-2.6` profiles were missing from `dev/deps`.

This patch updates these scripts to include all four Hadoop profiles defined in our POM.

### CODE REVIEW

File: dev/run-tests.py [hunk 1]
```diff
@@ -295,13 +295,14 @@ def exec_sbt(sbt_args=()):
...
 def get_hadoop_profiles(hadoop_version):
     """
-    For the given Hadoop version tag, return a list of SBT profile flags for
+    For the given Hadoop version tag, return a list of Maven/SBT profile flags for
     building and testing against that Hadoop version.
     """
 
     sbt_maven_hadoop_profiles = {
         "hadoop2.2": ["-Pyarn", "-Phadoop-2.2"],
-        "hadoop2.3": ["-Pyarn", "-Phadoop-2.3", "-Dhadoop.version=2.3.0"],
```
Comment (JoshRosen) [inline dev/run-tests.py]: Setting `hadoop.version` is unnecessary here; it's already set in the profile.

File: dev/run-tests.py [hunk 2]
```diff
@@ -295,13 +295,14 @@ def exec_sbt(sbt_args=()):
 
 def get_hadoop_profiles(hadoop_version):
     """
-    For the given Hadoop version tag, return a list of SBT profile flags for
+    For the given Hadoop version tag, return a list of Maven/SBT profile flags for
```
Comment (JoshRosen) [inline dev/run-tests.py]: These flags are used in both Maven and SBT, hence this clarification (to see this, grep the file for usages of `get_hadoop_profiles`).

**Comment by yhuai** [comment]
LGTM

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
This MiMa failure looks interesting; I wonder if we had a non-explicit return type which is affected by the Hadoop version...

**Comment by JoshRosen** [comment]
```
[info] spark-sql: found 2 potential binary incompatibilities (filtered 527)
[error]  * method leafFiles()scala.collection.mutable.Map in class org.apache.spark.sql.sources.HadoopFsRelation#FileStatusCache has now a different result type; was: scala.collection.mutable.Map, is now: scala.collection.mutable.LinkedHashMap
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.apache.spark.sql.sources.HadoopFsRelation#FileStatusCache.leafFiles")
[error]  * method leafFiles_=(scala.collection.mutable.Map)Unit in class org.apache.spark.sql.sources.HadoopFsRelation#FileStatusCache's type has changed; was (scala.collection.mutable.Map)Unit, is now: (scala.collection.mutable.LinkedHashMap)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.sql.sources.HadoopFsRelation#FileStatusCache.leafFiles_=")
[warn] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala:388: method 
```

**Comment by JoshRosen** [comment]
It occurred to me that we should probably also add these profiles to the `dev/deps` lists as well, esp. since the `hadoop-2.6` profile has some dependency overrides. I might as well do that as part of this PR to collect all of the Hadoop profile parity changes in one place.

**Comment by JoshRosen** [comment]
The test failures are unrelated, so I'm going to merge this into master in order to unblock some other fixes to test infra.

**Pull request closed without merging**